### PR TITLE
chore(ci): Move most CI checks to merge queue

### PR DESCRIPTION
**Pull request by neuronull**
### TLDR of behavior changes you need to know

- Each commit to a PR will now only run a subset of the tests. Once those pass (and required review approvals received), the PR can enter the merge queue.
    - Instead of relying on CI detect issues, it will be more efficient for developers to run some local tests that they think would make sense to run based on their changes. For example, integration tests for your integration, k8s e2e tests, mac or windows tests. If there are challenges to running those locally, you can manually trigger these with a comment to the PR, see the next section for details on that.
- PRs enter the merge queue instead of merging directly to master. The merge queue runs the rest of the required tests. If it passes, it is merged into master, otherwise the PR remains open for the author to address whatever failed.
- When the PR is merged into master- there is currently no option to manually set the squashed commit message. The best option we have right now, is it will automatically use the PR title as the commit title, and the opening PR comment as the description of the commit. 
    - This means developers should write a concise quality opening PR comment that they are ok with being the commit description for their PR merging into master. Apparently there is no limit to the character limits on commit descriptions, so we will need to be vigilant about providing extended context explained in a second comment in the PR.

### Overview

1. Master merge queue
    - Most workflows that did not frequently detect failures were moved to the merge queue.
        - The workflow that still runs on each PR commit is `Test Suite`, which has the linux x86 unit tests, all the cargo checks, and VRL tests.
            - Currently the Regression workflow(s) also run on each PR commit, but those are targeted to be included in the merge queue in a follow-up PR.
    - The merge queue run occurs when your PR has passed the `Test Suite`, and you click "Merge when ready" (or have enabled auto merge).
    - If there are no PRs ahead of you in the queue, yours will trigger shortly, otherwise it gets queued up, see the screenshots below.  Currently it takes roughly 1 hour for the merge queue checks to complete.
    - Once it passes the merge queue checks, it is merged into master and the PR closes.
    - If the merge queue checks fail, the PR is removed from the queue, and the details of the failed CI run are available in the PR to view why.
    
2. Improved file changes detection
    - Re-usable workflow `changes.yml` provides more extensive outputs for detecting what files changed, most notably these are utilized to decide whether the k8s workflow or any of the specific integration tests should be run. For example, if your PR makes a change in only the `amqp` source, the file detection logic will result in the `amqp` integration tests running as part of the merge queue checks, but not the other integration tests.

3. Replaces PR label CI triggers with more extensive set of PR comment triggers.
    - Previously we added labels like "ci-condition: integration tests enable" , to PRs in order to get some tests to run that weren't part of the regular checks. This has a downside of always running those checks on each commit thereafter.
    - This was replaced by the ability to add comments to a PR that will run a workflow on demand. The available workflows and their triggers are listed below.
        - The integration ones are a special case. After the "/ci-deploy-integration" is matched , any matching keyword for one of the integrations, will add that to the matrix set that gets run. See 'integration-comment.yml' for examples. Note that 'all' is a key word which will run all of the integrations in the matrix set.
    
| Test | Comment text |
| --- | ----------- |
| *all  | /ci-run-all |
| Cross  | /ci-run-cross |
| CLI - Linux | /ci-run-cli |
| Unit - Windows | /ci-run-unit-windows |
| Unit - Mac | /ci-run-unit-mac |
| Misc | /ci-run-misc |
| K8s E2E  | /ci-run-k8s |
| Install.sh  | /ci-run-install |
| Environment | /ci-run-environment |
| Regression  | /ci-run-regression |
| Integration  | /ci-run-integration* |

### TODO follow up PRs

- [x] https://github.com/vectordotdev/vector/pull/17399
    - [x] Combine the two workflows into one, since the merge queue is considered trusted.
    - [x] Trigger regression workflow on comment '/ci-deploy-regression'
    - [x] Add changes filter to regression workflow- only run if source was changed

- [x] https://github.com/vectordotdev/vector/pull/17416
- [ ] Move the file detection logic for int tests to the `test.yaml` for the specific integration. Add a vdev subcommand that reads it and spits out the paths. Parse that in CI.
- [x] add a `/ci-run-all`

### Future enhancements

- [ ] Investigate using configuration variables at the repository level for repeated env vars in workflows, accessed through the GH Context
- [ ] Clean up the env vars as there are some copy pasted into workflows that probably don't need them (though they aren't harming anything by being there)
- [ ] The integration workflow dispatch trigger requires an input for test name. It would be nice to have that accept "all" and run all the integration tests on demand.

### Master Merge Queue Suite workflow

***

Screenshot 2023-05-05 at 1 34 19 PM

### Test Suite workflow

***

Screenshot 2023-05-05 at 1 34 32 PM

### Merge queue in action

***
Screenshot 2023-05-09 at 9 52 11 AM

Screenshot 2023-05-09 at 9 52 30 AM

Screenshot 2023-05-05 at 3 39 29 PM

Screenshot 2023-05-11 at 11 07 30 AM

### main branch required checks

***

Screenshot 2023-05-11 at 11 07 53 AM

### CODE REVIEW

File: .github/workflows/master_merge_queue.yml [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+    with:
+      base_ref: ${{ github.event.merge_group.base_ref }}
+      head_ref: ${{ github.event.merge_group.head_ref }}
+    secrets: inherit
+
+  test-cli:
+    uses: ./.github/workflows/cli.yml
+    needs: changes
+    if: needs.changes.outputs.source == 'true'
+    secrets: inherit
```
Comment (spencergilbert) [inline .github/workflows/master_merge_queue.yml:56]:
```suggestion
    secrets: inherit
    if: needs.changes.outputs.source == 'true'
```

These different jobs having different orders for the options are hurting me 😂

**Review by spencergilbert — COMMENTED** [review]
As a minor complaint I find the `/ci-deploy-*` naming for the comment triggers odd - primarily the use of "deploy." I'd consider a different verb (run, exec, ...).

File: .github/workflows/changes.yml
[the remark this answers is not in the document]
Comment (neuronull) [reply .github/workflows/changes.yml:1]: Ah yes, I mentally had that as a TODO but failed to jot it down on the description.
Wherever we have new integration tests for a new component, I can add some documentation on required changes for getting them incorporated into CI correctly.

File: .github/workflows/integration-test.yml [hunk 2]
```diff
@@ -1,34 +1,28 @@
-name: Integration Test Suite
+# This workflow is used to run an integration test.
+# The most common use case is that it is triggered by another workflow,
+# such as the Master Merge Queue Suite, or the Integration Comment.
+#
+# It can also be triggered on manual dispatch in CI however.
+# In that use case, an input for the test name needs to be provided.
+# TODO: check if the input is "all" , and run all, without a timeout?
```
[the remark this answers is not in the document]
Comment (neuronull) [reply .github/workflows/integration-test.yml:7]: 🤔 I think when I wrote that I was thinking that if we ran all the integration tests together, it could imply overreaching the 30 min (or whatever we settle on) timeout.

File: .github/workflows/master_merge_queue.yml (hunk 1 above)
Comment (neuronull) [reply .github/workflows/master_merge_queue.yml:56]: 😂 I can fix that up.

File: .github/workflows/changes.yml [hunk 3]
```diff
@@ -0,0 +1,334 @@
...
+            - "vdev/**"
+          docker:
+            - 'distribution/docker/**'
+            - "vdev/**"
+          install:
+            - ".github/workflows/install-sh.yml"
+            - "distribution/install.sh"
+          k8s:
+            - "src/sources/kubernetes_logs/**"
+          amqp:
```
Comment (neuronull) [inline .github/workflows/changes.yml:201]: For the integration tests, it struck me that we might want to check for changes in common places like 'lib/' or something like that, and run all of them.
Wanted to get team's thoughts on that as I'm sure there would be some cases where that wouldn't be necessary, but figured "core" code utilized everywhere might benefit from that.

**Comment by neuronull** [comment]
> As a minor complaint I find the `/ci-deploy-*` naming for the comment triggers odd - primarily the use of "deploy." I'd consider a different verb (run, exec, ...).

Yeah definitely open to changing that. I like `/ci-run-<...>`better than deploy too.

File: .github/workflows/integration-comment.yml [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+      if: ${{ matrix.run.if }}
+      test_name: ${{ matrix.run.test_name }}
+    needs: prep-pr
+    secrets: inherit
+    strategy:
+      fail-fast: false
+      matrix:
+        run:
+          - test_name: 'amqp'
+            if: ${{ contains(github.event.comment.body, 'amqp') || contains(github.event.comment.body, 'all') }}
```
[the remark this answers is not in the document]
Comment (neuronull) [reply .github/workflows/integration-comment.yml:76]: That was a happy discovery for sure.

File: .github/workflows/component_features.yml [hunk 5]
```diff
@@ -0,0 +1,69 @@
+name: Component Features - Linux
+
+on:
+  workflow_call:
```
[the remark this answers is not in the document]
Comment (neuronull) [reply .github/workflows/component_features.yml:4]: I did consider that. The reasons I organized it this way and IMO the advantages are:
 - It is easier to know what is in the merge queue, because we aren't allowed to have workflows in directories, they all have to be flat in the .workflows dir, now we can see from the single master_merge_queue.yml file, what all is in the merge queue and that file itself is (mostly, with exception of the int tests), not complicated.
 - I think it is easier to click on the one workflow run 'Master Merge Queue Suite' and all the jobs there, rather than have to open a new tab to view each of them.
 
 I realize that is kind of opinionated so I'm def open to differing thoughts.

File: .github/workflows/integration-test.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (spencergilbert) [reply .github/workflows/integration-test.yml:7]: I don't think it _should_ make a difference. The timeout is per job, which in this case more integration tests are just more parallel jobs, so the same timeout _should_ work?

File: .github/workflows/install-sh.yml [hunk 6]
```diff
@@ -22,25 +17,9 @@ jobs:
...
-    needs:
-      - sync-install
+    needs: sync-install
     runs-on: ubuntu-20.04
     steps:
       - run: sudo apt-get install --yes curl bc
       - run:  curl --proto '=https' --tlsv1.2 -sSf https://sh.vector.dev | bash -s -- -y
       - run: ~/.vector/bin/vector --version
-
-  install-shell-failure:
```
[the remark this answers is not in the document]
Comment (spencergilbert) [reply .github/workflows/install-sh.yml:33]: _and also we should start using DD monitors anyways_

File: .github/workflows/integration-test.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (neuronull) [reply .github/workflows/integration-test.yml:7]: Ah, the critical information here is that this workflow is not the matrix.
It is a single run. And so the enhancement here would be running `make test-integration-all` in a single job.

File: .github/workflows/baseline.yml [hunk 7]
```diff
@@ -1,8 +1,8 @@
-# Executes various builds of vector to time the results in order to track build times.
+# Executes various builds of vector to time the results in order to track compilation times.
```
[the remark this answers is not in the document]
Comment (neuronull) [reply .github/workflows/baseline.yml:1]: ah, yes that was an oversight.

File: .github/workflows/integration-test.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (spencergilbert) [reply .github/workflows/integration-test.yml:7]: Oh, running serially? I did miss that.

File: .github/workflows/changes.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (neuronull) [reply .github/workflows/changes.yml:201]: I agree.. I did see they have an option to define them in a separate file. It shouldn't be hard to add that.
Though, from what I'm seeing it is a single file to read from 🤔 (as opposed to a file-per key)

File: .github/workflows/changes.yml [hunk 8]
```diff
@@ -0,0 +1,334 @@
...
+            - "src/internal_events/datadog_*"
+            - "src/sources/datadog_agent/**"
+            - "src/sinks/datadog/**"
+            - "src/sinks/datadog_archives.rs"
+            - "src/sinks/util/**"
+          docker-logs:
+            - "src/docker.rs"
+            - "src/internal_events/docker_logs.rs"
+            - "src/sources/docker_logs/**"
+            - "src/sources/util/**"
```
[the remark this answers is not in the document]
Comment (neuronull) [reply .github/workflows/changes.yml:242]: I too had some concerns about this (https://github.com/vectordotdev/vector/pull/17340#discussion_r1188945939)

Apart from any recommendations to run all int tests on specific areas, I'll make the behavior changes you suggested.

It also adds motivation to implement the changes to re use the same build for all the int tests.

File: .github/workflows/integration-test.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (neuronull) [reply .github/workflows/integration-test.yml:7]: The matrix is defined in the master merge queue workflow, and this one is called into run each individually. So technically they aren't serial.

File: .github/workflows/integration-comment.yml [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+#
+#     /ci-deploy-integration amqp
+#
+# 2. Run three specific integration tests:
+#
+#     /ci-deploy-integration dnstap redis amqp
+#
+# 3. Run all integration tests:
+#
+#     /ci-deploy-integration all
```
[the remark this answers is not in the document]
Comment (neuronull) [reply .github/workflows/integration-comment.yml:19]: One thing to be aware of, this won't affect people generally, but the nature of this triggering off of issue comments is that, every workflow we have that does this detection, gets logged in the GHA view as skipped if the comment doesn't have the matching text. For example, someone posts a regular comment on a PR saying, "bravo!" , if you then go look at the CI view, you will see about 5 or 6 runs show up as "skipped".

I wasn't able to find a way around that but it seems harmless other than being a bit ugly. I considered that the PR labels doesn't do this, but I think the benefits of per comment triggers outweigh this downside as most people don't look at the GHA view too often. Perhaps impacts the gardener the most.

[the remark this answers is not in the document]
Comment (spencergilbert) [reply .github/workflows/integration-comment.yml:19]: Mhmm, I assume you can get around it by having a proper GH app?

File: .github/workflows/integration-comment.yml
[the remark this answers is not in the document]
Comment (neuronull) [reply .github/workflows/integration-comment.yml:1]: I definitely will add documentation to the areas we have it.

Extending vdev to aid in a new component creation could be interesting, I like the spirit of it. Though would consider that a follow on enhancement that is tangential to the main work here.

[the remark this answers is not in the document]
Comment (spencergilbert) [reply .github/workflows/integration-comment.yml:1]: Yep, 100%

File: .github/workflows/changes.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (neuronull) [reply .github/workflows/changes.yml:201]: Though I suppose could split that out into different jobs.

[the remark this answers is not in the document]
Comment (jszwedko) [reply .github/workflows/changes.yml:201]: You could also add a subcommand to `vdev integration` that spits out the file paths by parsing the `test.yaml` files.

File: .github/workflows/component_features.yml (hunk 5 above)
[the remark this answers is not in the document]
Comment (jszwedko) [reply .github/workflows/component_features.yml:4]: Gotcha, yeah, those seem like nice benefits so I'm happy to go with it. I was just curious if you considered both ways.

File: .github/workflows/changes.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (neuronull) [reply .github/workflows/changes.yml:201]: Oh I do like that better... Ok, I think that is a good candidate for a follow-up task.

File: .github/workflows/integration-comment.yml [hunk 10]
```diff
@@ -1,22 +1,22 @@
 # This workflow runs one or more integration tests triggered by a comment in a PR.
-# The comment include '/ci-deploy-integration'.
-# Then, any if the integration names will trigger that specific integration.
+# The comment include '/ci-run-integration'.
+# Then, any if the integration names will triger that specific integration.
```
Comment (spencergilbert) [inline .github/workflows/integration-comment.yml]:
```suggestion
# Then, any if the integration names will trigger that specific integration.
```

Comment (neuronull) [reply .github/workflows/integration-comment.yml]: huh, this is outdated

Comment (spencergilbert) [reply .github/workflows/integration-comment.yml]: 🤷 , gremlins

**Review by spencergilbert — APPROVED** [review]

**Comment by neuronull** [comment]
An FYI- I've made changes for all of the feedback so far (apart from those identified as follow up tasks) , but I'm still testing out some of the changes made in the playground. The GH outages is delaying that.

File: .github/workflows/integration-comment.yml (hunk 9 above)
[the remark this answers is not in the document]
Comment (neuronull) [reply .github/workflows/integration-comment.yml]: Hmm I am unknowledgeable about GH apps ... I'll have to read up on that to understand what that might entail and how it solves the problem.

File: .github/workflows/k8s_e2e.yml [hunk 11]
```diff
@@ -57,6 +57,7 @@ jobs:
       CARGO_PROFILE_RELEASE_CODEGEN_UNITS: 256
       CARGO_INCREMENTAL: 0
     steps:
+<<<<<<< HEAD
```
Comment (jszwedko) [inline .github/workflows/k8s_e2e.yml:60]: 👀

File: .github/workflows/environment.yml [hunk 12]
```diff
@@ -15,8 +15,30 @@ jobs:
   publish-new-environment:
     runs-on: ubuntu-20.04
     steps:
-      - name: Checkout
+      - name: (PR comment) Get PR branch
+        if: ${{ github.event_name == 'issue_comment' }}
```
Comment (jszwedko) [inline .github/workflows/environment.yml:19]: Is this right? It seems like there is no `issue_comment` trigger anymore. Same applies to `install-sh.yml`.

**Review by jszwedko — COMMENTED** [review]
A couple of more comments!

**Review by jszwedko — COMMENTED** [review]
A couple of more comments!

File: .github/workflows/environment.yml (hunk 12 above)
Comment (neuronull) [reply .github/workflows/environment.yml:19]: Ah yeah that isn't obvious but when the trigger is `workflow_call` , the github event context is the same one as the called workflow. So this and the workflows where we call into it from another, will still have this logic because the caller workflow context event_name is `issue_comment`.

I first tried a more "refined" (or so I thought) approach where the comment trigger workflow itself would set the pending status, and final status as well, after calling the workflow. But it became messy really quick, and the status check that shows up on the PR would be "Comment Trigger", unless we passed in the name as input param to `workflow_call` and that just seemed ridiculous.
Given that our main goal with this was to reduce the number of workflows that specifically activated `on` `issue_comment`, the current design seemed the best approach.

Comment (jszwedko) [reply .github/workflows/environment.yml:19]: Ah! Yes, I didn't realize the github event context is the same. 👍
